Code monitoring: Implement in-context test action

novoselrok opened the pull request:
Prepares backend for #17036

quinnkeast commented:
Looking great! Two comments:

- [ ] The logo needs to be replaced at the source, as it currently uses the old version
- [ ] I don't know if this is addressable, but in the email preview (in Hey), there's a missing space in the notice part:

<img width="338" alt="image" src="https://user-images.githubusercontent.com/6304497/104290095-3680e800-54ba-11eb-8d65-0bee1ae6d7c6.png">

quinnkeast commented:
Updated docs in #17196 to align with this new functionality.

novoselrok commented:
@quinnkeast I updated the logo in the email and added a non-breakable space between the sentences in the test email. Can you please check again if the issue persists in the Hey client?

quinnkeast commented:
> @quinnkeast I updated the logo in the email and added a non-breakable space between the sentences in the test email. Can you please check again if the issue persists in the Hey client?

Perfect!

stefanhengl commented:
It may be better to split this PR into frontend and backend changes. The changes in the backend should be independent of the frontend, aren't they? It would be easier to review and rollback. IMO it is best to keep things separate unless there is a good reason to roll out frontend and backend together.

novoselrok commented:
> It may be better to split this PR into frontend and backend changes. The changes in the backend should be independent of the frontend, aren't they? It would be easier to review and rollback. IMO it is best to keep things separate unless there is a good reason to roll out frontend and backend together.

Ok, I'll split the client changes from this PR and create a new PR for them.

novoselrok commented:
I've split the client changes here: https://github.com/sourcegraph/sourcegraph/pull/17247

stefanhengl commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -839,6 +839,10 @@ type Mutation {
+    """
+    Triggers a test email for a code monitor action.
+    """
+    triggerTestEmailAction(user: ID!, description: String!): EmptyResponse!
```
Edit: Just tested locally, we cannot create an action without a trigger which means we already have a description that we can use in the test email. 

@quinnkeast  The design of the test email doesn't have a subject line, is this intentional?

quinnkeast replied:
@stefanhengl looks like something happened in Figma. There's a subject line in the implementation: `Test: [New event] Title here`. I'll add that back to Figma.

novoselrok replied:
> This is good enough for now but as soon as we permit more than one email you would have to update the API again. We should allow a list here or maybe even an object like MonitorEmailInput. WDYT?

I think using the `MonitorEmailInput` is a good idea, because we might want to use other fields for testing eventually, e.g.: email priority and header.

stefanhengl commented on enterprise/internal/codemonitors/resolvers/resolvers_test.go:
```diff
@@ -1060,3 +1061,44 @@ query($userName: String!, $actionCursor:String!, $actionEventCursor:String!){
+	}
+
+	ctx := backend.WithAuthzBypass(context.Background())
+	dbtesting.SetupGlobalTestDB(t)
```
This is not a blocker. Can you think of a way to test this without setting up a test db? This op is very expensive. Is there another test we can merge this with? Or maybe we can mock the auth check, or ...?

stefanhengl reviewed (approved):
Awesome!
Left some comments, no blockers.

I wonder whether the email package should move to the same level as the resolvers package since now we send emails from the background workers as well as from the resolvers.

novoselrok commented:
> I wonder whether the email package should move to the same level as the resolvers package since now we send emails from the background workers as well as from the resolvers.

@stefanhengl I agree, I moved the email package one level up. I also addressed the rest of the comments, feel free to check again.

stefanhengl commented:
LGTM!
